# Домашнє завдання до теми "Розширені методи обробки даних "

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [5]:
import pandas as pd
df = pd.read_csv('/content/drive/MyDrive/supermarket_sales.csv')
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [6]:
df.sort_values('Unit price', ascending=False)[:5]

Invoice ID Branch       City Customer type Gender         Product line  \
983  148-41-7930      C  Naypyitaw        Normal   Male    Health and beauty   
122  219-22-9386      B   Mandalay        Member   Male    Sports and travel   
930  641-62-7288      B   Mandalay        Normal   Male   Home and lifestyle   
494  437-53-3084      B   Mandalay        Normal   Male  Fashion accessories   
283  667-92-0055      A     Yangon        Member   Male    Health and beauty   

     Unit price  Quantity  Tax 5%    Total       Date   Time      Payment  \
983       99.96         7  34.986  734.706  1/23/2019  10:33         Cash   
122       99.96         9  44.982  944.622   3/9/2019  17:26  Credit card   
930       99.92         6  29.976  629.496  3/24/2019  13:33      Ewallet   
494       99.89         2   9.989  209.769  2/26/2019  11:48      Ewallet   
283       99.83         6  29.949  628.929   3/4/2019  15:02      Ewallet   

       cogs  gross margin percentage  gross income  Rating  
983  699.72                 4.761905        34.986     6.1  
122  899.64                 4.761905        44.982     4.2  
930  599.52                 4.761905        29.976     7.1  
494  199.78                 4.761905         9.989     7.1  
283  598.98                 4.761905        29.949     8.5

2. Знайдіть інвойс (рядок в даних) в категорії товарів Electronic accessories, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [7]:
df[df['Product line']=='Electronic accessories'].sort_values(['Unit price', 'Quantity'], ascending=[False, False])[:1]

Invoice ID Branch      City Customer type  Gender  \
209  817-69-8206      B  Mandalay        Normal  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
209  Electronic accessories       99.73         9  44.8785  942.4485   

         Date   Time      Payment    cogs  gross margin percentage  \
209  3/2/2019  19:42  Credit card  897.57                 4.761905   

     gross income  Rating  
209       44.8785     6.5

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу: високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 до 8 не включаючи - то він середній (`Medium`)
- інакше - низький (`Low`).

In [12]:
def get_rating_category(val):
  if val >= 8:
    return "High"
  elif val >= 6 and val < 8:
    return "Medium"
  else:
    return "Low"

In [14]:
df['Rating Category'] = df.Rating.apply(get_rating_category)
df['Rating Category']

0        High
1        High
2      Medium
3        High
4         Low
        ...  
995    Medium
996       Low
997    Medium
998       Low
999    Medium
Name: Rating Category, Length: 1000, dtype: object

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [15]:
df['Rating Category'].value_counts(normalize=True)

Medium    0.356
High      0.329
Low       0.315
Name: Rating Category, dtype: float64

4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [17]:
time = '13:13'

In [18]:
def get_hour(time):
    return int(time.split(':')[0])

In [19]:
get_hour(time)

13

4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [20]:
df['Hour'] = df.Time.apply(get_hour)

4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [21]:
df['Hour'].value_counts()

19    113
13    103
15    102
10    101
18     93
11     90
12     89
14     83
16     77
20     75
17     74
Name: Hour, dtype: int64

Альтернативно:

In [22]:
df.groupby('Hour').size()

Hour
10    101
11     90
12     89
13    103
14     83
15    102
16     77
17     74
18     93
19    113
20     75
dtype: int64

Найбільше покупок здійснюється о 19:00, з високою активністю також о 13:00 та 15:00. Це логічно, адже ввечері люди зазвичай завершують робочий день, а обідня перерва дає можливість зробити швидкі покупки. Ці часові періоди є ключовими для бізнесу для планування персоналу, запасів і маркетингових акцій.

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [23]:
df.groupby('City')['Total'].mean()

City
Mandalay     319.872506
Naypyitaw    337.099715
Yangon       312.354031
Name: Total, dtype: float64

6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [24]:
df.groupby('City')['Total'].agg(['count', 'mean', 'min', 'max'])

count        mean      min      max
City                                          
Mandalay     332  319.872506  18.6375  1022.49
Naypyitaw    328  337.099715  10.6785  1042.65
Yangon       340  312.354031  12.6945  1039.29

Ці дані показують кількість покупок (count), середню (mean), мінімальну (min) та максимальну (max) суму чеків (Total) по кожному місту:

**Mandalay** має 332 покупки з середньою сумою чека 319.87, мінімальною сумою 18.64 і максимальною 1022.49.

**Naypyitaw** має 328 покупок, середня сума чека — 337.10, мінімальна — 10.68, максимальна — 1042.65.

**Yangon** має 340 покупок, з середньою сумою чека 312.35, мінімальною сумою 12.69 і максимальною 1039.29.

###**Висновки:**

**Naypyitaw** має найвищу середню суму чека, що може свідчити про вищу покупну спроможність клієнтів або популярність дорожчих товарів у цьому місті.

**Yangon** лідирує за кількістю покупок, але має трохи меншу середню суму чека, порівняно з **Naypyitaw**.

**Mandalay**, маючи середню кількість покупок, тримає середні показники за сумою чека.

###**Рекомендації:**

*   Аналізувати асортимент у **Naypyitaw**: Оскільки середня сума чека тут найвища, варто проаналізувати, які товари найбільше купують, і розглянути можливість розширення цих категорій товарів.

*   Стимулювання продажів у **Yangon**: Враховуючи найбільшу кількість покупок, можна запровадити програми лояльності, акції або знижки, щоб збільшити середню суму чека.

*   Оптимізація запасів у **Mandalay**: Потрібно переконатися, що запаси відповідають попиту, особливо на товари, що користуються найбільшою популярністю, зважаючи на стабільну кількість покупок і середні суми чеків.

*   Дослідження покупців: Зібрати додаткові дані про покупців у кожному місті, щоб зрозуміти їхні переваги та поведінку, що може допомогти у прийнятті рішень щодо ціноутворення, маркетингу та асортименту товарів.

Такий аналіз допоможе супермаркетам краще розуміти своїх клієнтів і оптимізувати свою бізнес-модель для максимізації прибутку.



7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [11]:
df.to_csv('/content/drive/MyDrive/supermarket_sales_enriched.csv', index=False)